src/adaptivespatialsampling.jl

```diff
@@ -0,0 +1,79 @@
+"""
+    AdaptiveSpatialSampling
+
+...
+
+**numpoints**, an Integer (def. 50), specifying the number of points to use.
+
+**α**, an AbstractFloat (def. 1.0), specifying ...
+"""
+Base.@kwdef mutable struct AdaptiveSpatialSampling{T <: Integer} <: BONRefiner
+    numpoints::T = 50
+    function AdaptiveSpatialSampling(numpoints)
+        if numpoints < one(numpoints)
+            throw(ArgumentError("You cannot have an AdaptiveSpatialSampling with fewer than one point"))
+        end
+        return new{typeof(numpoints)}(numpoints)
+    end
+end
+
+function _generate!(coords::Vector{CartesianIndex}, pool::Vector{CartesianIndex}, sampler::AdaptiveSpatialSampling, uncertainty::Matrix{T}) where {T<:AbstractFloat}
+
+    # Distance matrix (inlined)
+    d = zeros(Float64, Int((sampler.numpoints * (sampler.numpoints - 1)) / 2))
+
+    # Start with the point with maximum entropy
+    imax = last(findmax([uncertainty[i] for i in pool]))
+    # Add it to the stack
+    coords[1] = popat!(pool, imax)
+
+    best_score = 0.0
+    best_s = 1
+    
+    for i in 2:sampler.numpoints
+        for (ci, cs) in enumerate(pool)
+            coords[i] = cs
+            # Distance update
+            start_from = Int((i - 1) * (i - 2) / 2) + 1
+            end_at = start_from + Int(i - 2)
+            d_positions = start_from:end_at
+            for ti in 1:(i-1)
+                d[d_positions[ti]] = D(cs, coords[ti])
+            end
+            # Get the score
+            score = uncertainty[cs] + sqrt(log(i)) * h(d[1:end_at], 1.0, 0.5)
+            if score > best_score
+                best_score = score
+                best_s = ci
+            end
+        end
+        coords[i] = popat!(pool, best_s)
+    end
+    return coords
+end
+
+function H(threshold::T, uncertainty::Matrix{T}) where {T<:AbstractFloat}
```

Comment:
If the functions below are not outward facing, should we prefix with an underscore?

Reply:
definitely, will fix